File: Unidad I/PAC II 2019/Codigos/3-LinkedList-Queue/JS/Queue.js
function Node(value) {
    this.value = value;
    this.next = null; //es a lo que apunta al nodo
}

function LinkedList() {
    this.first = null;
    this.add = LinkedListAdd;
    this.print = LinkedListPrint;
    this.addInPosition = LinkedListAddInPosition;
    this.getLast = LinkedListGetLast;
    this.getFirst = LinkedListGetFirst;
    this.delete = LinkedListDelete;
}

function LinkedListAdd(value) {
    //si no existe uno primero lo agrega
    if (!this.first) {
        this.first = new Node(value);
    } //cuando ya existe un primero buscar el ultimo
    //Y agregar el nuevo nodo despues
    else {
        current = this.first;
        while (current.next) {
            current = current.next;
        }
        current.next = new Node(value);
    }
}

function LinkedListDelete(value) {
    prev = null;
    current = this.first;
    if (this.first == null) {
        return false;
    } else {
        while (current.value != value && current.next != null) {
            prev = current;
            current = current.next;
        }

        if (current.value == value) {
            if (current == self.first) {
                if (current.next == null) {
                    this.first = null;
                } else {
                    this.first = current.next;
                }
            } else {
                if (current.next == null) {
                    prev.next = null;
                } else {
                    prev.next = current.next;
                }

            }

        }

    }

}

function LinkedListPrint() {
    current = this.first;
    while (current.next) {
        console.log(current.value);
        current = current.next;
    }
    console.log(current.value);
}

function LinkedListAddInPosition(value, n) {
    count = 0;

    //agrega en el inicio el value 
    if (n == 0) {
        queue = this.first;
        this.first = new Node(value);
        this.first.next = queue;

        return true;
    } else if (n > 0) { //agrega despues del inicio el value
        current = this.first;
        while (current.next) {
            centinela = current; //se necesita guardar la posicion donde se "corta"
            current = current.next;
            count++;
            if (count == n) {

                queue = current;
                current = new Node(value);
                current.next = queue;
                centinela.next = current; //Se usa el centinela para obtener la posicion cortada
                return true;
            }
        }
    }

    return false;
}

function LinkedListGetFirst() {

    return this.first;
}

function LinkedListGetLast() {
    last = this.first;
    while (last.next) {
        last = last.next;
    }
    return last;
}

/*
list = new LinkedList();
list.add("1");
list.add("Hola Mundo");
list.add(2);

 //Lista: "1","2","Hola Mundo","2"
list.addInPosition("6",0);  

console.log("El valor del primer nodo es: ", list.getFirst().value);
console.log("El valor del ultimo nodo es: ", list.getLast().value);



list.print();
//Lista: "1","2","Hola Mundo","2"
list.addInPosition("6", 2);
list.print();*/

File: Unidad I/PAC II 2019/Codigos/3-LinkedList-Queue/Python/Queue.py
#-*- coding:utf8 -*-
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

class LinkedList:
    def __init__(self):
        self.first = None

    def add(self, value):
        if not self.first:
            self.first = Node(value)
            return True
        else: 
            current = self.first
            while(current.next):
                current = current.next
            current.next = Node(value)
            return True
        return False

    def addInPosition(self, value, position):
        count = 0
        if position == 0:
            queue = self.first
            self.first = Node(value)
            self.first.next = queue

            return True
        else:
            if position > 0:
                current = self.first
                before = self.first
                while current.next:
                    current = current.next
                    count = count +1
                    if count == position:
                        before.next = Node(value)
                        before.next.next = current
                        return True
                    before = before.next
        return False
    
    def delete(self, value):
        prev = None
        current = self.first
        if self.first is None:
            return False
        else:
            while (current.value != value and current.next is not None):
                prev = current
                current = current.next
            if(current.value == value):
                if current == self.first:
                    if current.next is None:
                        self.first = None
                    else:
                        self.first = current.next
                else:
                    if current.next is None:
                        prev.next = None
                    else:
                        prev.next = current.next
            else:
                return False

    def getFirst(self):
        return self.first

    def getLast(self):
        last = self.first
        while last.next:
            last = last.next
        return last

    def printLL(self):
        current = self.first
        while current.next:
            print(current.value)
            current = current.next
        print(current.value)

